crates/transport/src/ws.rs

```diff
@@ -136,32 +144,83 @@ impl KeepAliveShutdown for std::sync::Arc<tokio::sync::Notify> {
     }
 }
 
+pub struct CloseWebsocketHandle {
+    sender: tokio::sync::mpsc::Sender<WsCloseFrame>,
+}
+
+// Todo: maybe we should consider using 1014 to indicates Bad Gateway? https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent/code
+// Note: Never sends 1005 and 1006 manually, as specified in RFC6455, section 7.4.1
+impl CloseWebsocketHandle {
+    pub async fn normal_close(self) -> Result<(), CloseError> {
+        self.sender
+            .send(WsCloseFrame {
+                code: 1000,
+                message: String::new(),
+            })
+            .await
+            .map_err(CloseError)
+    }
+
+    pub async fn server_error(self, message: &str) -> Result<(), CloseError> {
+        self.sender
+            .send(WsCloseFrame {
+                code: 1011,
+                message: message.to_owned(),
+            })
+            .await
+            .map_err(CloseError)
+    }
+}
+
+#[derive(Debug)]
+pub struct CloseError(tokio::sync::mpsc::error::SendError<WsCloseFrame>);
+
+impl Display for CloseError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
```

Comment:
nitpick: I don’t think we need to keep around the original error.
```suggestion
#[derive(Debug)]
pub struct CloseError;

impl Display for CloseError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "WebSocket already closed")
    }
}
```